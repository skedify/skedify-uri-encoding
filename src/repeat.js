export default function repeat(string, count) {
  if (string.repeat && typeof string.repeat === 'function') {
    return string.repeat(count);
  }

  let counter = Number(count);
  let newString = string;

  if (isNaN(counter)) {
    counter = 0;
  }

  if (counter < 0) {
    throw new RangeError('repeat count must be non-negative');
  }

  if (counter === Infinity) {
    throw new RangeError('repeat count must be less than infinity');
  }

  counter = Math.floor(counter);

  if (newString.length === 0 || counter === 0) {
    return '';
  }

  // Ensuring count is a 31-bit integer allows us to heavily optimize the
  // main part. But anyway, most current (August 2014) browsers can't handle
  // strings 1 << 28 chars or longer, so:
  if (newString.length * counter >= 1 << 28) {
    throw new RangeError('repeat count must not overflow maximum string size');
  }

  let rpt = '';

  /* eslint-disable-next-line no-constant-condition */
  while (true) {
    /* eslint-disable-next-line no-bitwise */
    if ((counter & 1) === 1) {
      rpt += newString;
    }

    /* eslint-disable-next-line no-bitwise */
    counter >>>= 1;

    if (counter === 0) {
      break;
    }

    newString += newString;
  }

  return rpt;
}
