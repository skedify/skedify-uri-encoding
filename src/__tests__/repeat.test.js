import repeat from '../repeat';

describe('repeat (when String.repeat exist)', () => {

  test('should return aWord five times', () => {
    expect(repeat('aWord', 5)).toMatchSnapshot();
  });

  test('should return empty string when empty string given', () => {
    expect(repeat('', 5)).toMatchSnapshot();
  });

  test('should return empty string when count is null', () => {
    expect(repeat('aWord', 0)).toMatchSnapshot();
  });

});

const aString = new String('aWord');
const aEmptyString = new String('');

aString.repeat = undefined;
aEmptyString.repeat = undefined;

describe('repeat (when String.repeat does not exist)', () => {

  test('should return aWord five times', () => {
    expect(repeat(aString, 5)).toMatchSnapshot();
  });

  test('should return empty string when empty string given', () => {
    expect(repeat(aEmptyString, 5)).toMatchSnapshot();
  });

  test('should return empty string when count is null', () => {
    expect(repeat(aString, 0)).toMatchSnapshot();
  });

  test('should throw error when count is negative', () => {
    expect(() => repeat(aString, -2)).toThrowErrorMatchingSnapshot();
  });

  test('should throw error when count is Infinity', () => {
    expect(() => repeat(aString, Infinity)).toThrowErrorMatchingSnapshot();
  });
});
