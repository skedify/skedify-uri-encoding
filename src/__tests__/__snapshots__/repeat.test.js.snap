// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`repeat (when String.repeat does not exist) should return aWord five times 1`] = `"aWordaWordaWordaWordaWord"`;

exports[`repeat (when String.repeat does not exist) should return empty string when count is null 1`] = `""`;

exports[`repeat (when String.repeat does not exist) should return empty string when empty string given 1`] = `""`;

exports[`repeat (when String.repeat does not exist) should throw error when count is Infinity 1`] = `"repeat count must be less than infinity"`;

exports[`repeat (when String.repeat does not exist) should throw error when count is negative 1`] = `"repeat count must be non-negative"`;

exports[`repeat (when String.repeat exist) should return aWord five times 1`] = `"aWordaWordaWordaWordaWord"`;

exports[`repeat (when String.repeat exist) should return empty string when count is null 1`] = `""`;

exports[`repeat (when String.repeat exist) should return empty string when empty string given 1`] = `""`;
